p2p: Unify Send and Receive protocol versions

hebasto opened the pull request:
On master (6fef85bfa3cd7f76e83b8b57f9e4acd63eb664ec) `CNode` has two members to keep protocol version:
- `nRecvVersion` for received messages
- `nSendVersion` for messages to send

After exchanging with `VERSION` and `VERACK` messages via protocol version `INIT_PROTO_VERSION`, both nodes set `nRecvVersion` _and_ `nSendVersion` to _the same_ value which is the greatest common protocol version.

This PR:
- replaces two `CNode` members, `nRecvVersion` `nSendVersion`, with `m_greatest_common_version`
- removes duplicated getter and setter

There is no change in behavior on the P2P network.

MarcoFalke commented:
> The change in behavior: with this PR VERACK message is sent and received with the common protocol version rather than INIT_PROTO_VERSION.

How does this change behavior? verack has no payload, so serialization flags can't possibly change the serialization.

hebasto commented:
> How does this change behavior? verack has no payload, so serialization flags can't possibly change the serialization.

You are right. I did not mean change of external behavior. Bad wording. Mind help me re-word it?

laanwj commented:
If correct, this is a very nice simplification.

laanwj commented:
>   You are right. I did not mean change of external behavior. Bad wording. Mind help me re-word it?
Or just skip it?

I'd suggest removing that, and explicitly say that there is *no* change in behavior on the P2P network.

hebasto commented:
@laanwj 
> > You are right. I did not mean change of external behavior. Bad wording. Mind help me re-word it?
> > Or just skip it?
> 
> I'd suggest removing that, and explicitly say that there is _no_ change in behavior on the P2P network.

Done. Thank you for your suggestion.

hebasto committed:
p2p: Unify Send and Receive protocol versions

There is no change in behavior on the P2P network.

hebasto committed:
refactor: Rename local variable nSendVersion

hebasto committed:
p2p: Remove SetCommonVersion() from VERACK handler

SetCommonVersion() is already called from the VERSION message handler.
There is no change in behavior on the P2P network.

hebasto commented:
Rebased 25b93f13f08fba0a25c3480e158d131eea8795e5 -> 5a89c1792ed3658d145c9ed3353e3f596727c2fe ([pr17785.02](https://github.com/hebasto/bitcoin/commits/pr17785.02) -> [pr17785.03](https://github.com/hebasto/bitcoin/commits/pr17785.03)) due to the conflict with #19053.

hebasto commented:
Rebased 25b93f13f08fba0a25c3480e158d131eea8795e5 -> cc5dba8e1fd2586dee32e1d19e8867e5f6b03754 ([pr17785.02](https://github.com/hebasto/bitcoin/commits/pr17785.02) -> [pr17785.04](https://github.com/hebasto/bitcoin/commits/pr17785.04)) due to the conflict with #19053.

hebasto commented:
Rebased cc5dba8e1fd2586dee32e1d19e8867e5f6b03754 -> e268f6451fd2e5b57e8d7409c2853bea984a1ba9 ([pr17785.04](https://github.com/hebasto/bitcoin/commits/pr17785.04) -> [pr17785.05](https://github.com/hebasto/bitcoin/commits/pr17785.05)) due to the conflict with #14033.

hebasto commented:
@jnewbery Mind looking into this PR?

jnewbery commented:
Concept ACK. The version fields in `CNode` are confusing. After version negotiation, I think we should only need one field to track common version. A couple of comments:

- there's also an `nVersion` field. Can that also be merged?
- would it make sense for `CNode` or `CNodeState` to store a `CNetMsgMaker` rather than constructing it in every function it needs to use it in?

MarcoFalke commented:
Concept ACK

jnewbery commented:
> If our node is not upgraded, comparing to its peer, nVersion and m_greatest_common_version would differ.

Right, but there's nothing we can do with that information. If a node has a higher version than us, it means that it has some functionality that we're unaware of. We can't do anything with that information except perhaps report it in logs and in getpeerinfo. All peer logic in net_processing should be based on the highest common version.

hebasto committed:
p2p: Use the greatest common version in peer logic

hebasto commented:
Updated 7b12f25d34e7af6a518e17456a48be493282deb3 -> 659a90cb93cbe9a894cae461e0401ed32b25caf3 ([pr17785.06](https://github.com/hebasto/bitcoin/commits/pr17785.06) -> [pr17785.07](https://github.com/hebasto/bitcoin/commits/pr17785.07)):
- rebased due to the conflict with #19704
- implemented @jnewbery's [suggestion](https://github.com/bitcoin/bitcoin/pull/17785#issuecomment-679040954):
> All peer logic in net_processing should be based on the highest common version.

jnewbery commented:
utACK 659a90cb93cbe9a894cae461e0401ed32b25caf3. Nice tidy-up. Thanks hebasto!

hebasto commented:
Rebased 659a90cb93cbe9a894cae461e0401ed32b25caf3 -> dc56e956fb988003777a949350f08b7964bd3452 ([pr17785.07](https://github.com/hebasto/bitcoin/commits/pr17785.07) -> [pr17785.08](https://github.com/hebasto/bitcoin/commits/pr17785.08)) due to the conflict with #19668.

hebasto commented:
Rebased dc56e956fb988003777a949350f08b7964bd3452 -> e68912bc559197d3c82c581a1f64ec4c06f119da ([pr17785.08](https://github.com/hebasto/bitcoin/commits/pr17785.08) -> [pr17785.09](https://github.com/hebasto/bitcoin/commits/pr17785.09)) due to the conflict with #19067.

jnewbery commented:
utACK e68912bc559197d3c82c581a1f64ec4c06f119da

hebasto commented:
Rebased e68912bc559197d3c82c581a1f64ec4c06f119da -> 646ddf115a7d82d5b2e05452e25736fb65cf9f03 ([pr17785.09](https://github.com/hebasto/bitcoin/commits/pr17785.09) -> [pr17785.10](https://github.com/hebasto/bitcoin/commits/pr17785.10)) due to the conflict with #19724.

jnewbery commented:
utACK 646ddf115a7d82d5b2e05452e25736fb65cf9f03

(verified `git range-diff e68912bc55~4..e68912bc55 646ddf115a~4..646ddf115a`. This was a trivial rebase)

hebasto commented:
Rebased 646ddf115a7d82d5b2e05452e25736fb65cf9f03 -> ddefb5c0b759950942ac03f28c43b548af7b4033 ([pr17785.10](https://github.com/hebasto/bitcoin/commits/pr17785.10) -> [pr17785.11](https://github.com/hebasto/bitcoin/commits/pr17785.11)) due to the conflict with #19791.

benthecarman reviewed (approved):
utACK `ddefb5c`

jnewbery commented:
ACK ddefb5c0b759950942ac03f28c43b548af7b4033

Verified `git range-diff 646ddf115a~4..646ddf115a ddefb5c0b7~4..ddefb5c0b7`

hebasto commented:
@MarcoFalke @amiti @sipa Mind looking into this PR?

naumenkogs commented:
ACK ddefb5c0b759950942ac03f28c43b548af7b4033
good refactoring

amitiuttarwar reviewed (commented):
I've taken a look at the code. Overall it looks like a very nice cleanup. I just have one question I'd like to better understand (about wtxidrelay / verack message) before I'm ready to leave a proper ACK.

I wondering if after these changes we can get rid of the `CNode.nVersion` field entirely. I gave it a quick shot here: https://github.com/amitiuttarwar/bitcoin/commit/ef2b8620dcc4123a3f62be48943490f72c40cfaf. It compiles & I think it makes sense (maybe one clarification in a log), but haven't evaluated super closely. 

thanks for this cleanup! its a very nice simplification.

hebasto commented:
@amitiuttarwar 
> I wondering if after these changes we can get rid of the `CNode.nVersion` field entirely. I gave it a quick shot here: [amitiuttarwar@ef2b862](https://github.com/amitiuttarwar/bitcoin/commit/ef2b8620dcc4123a3f62be48943490f72c40cfaf). It compiles & I think it makes sense (maybe one clarification in a log), but haven't evaluated super closely.

@jnewbery [raised](https://github.com/bitcoin/bitcoin/pull/17785#issuecomment-656880630) the same question. This could be resolved in a follow-up PR.
I think we need smth like `bool CNode::VersionMsgAlreadyReceived()`.

hebasto commented on src/net_processing.cpp:
```diff
@@ -2406,10 +2406,10 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             PushNodeVersion(pfrom, m_connman, GetAdjustedTime());
 
         if (nVersion >= WTXID_RELAY_VERSION) {
-            m_connman.PushMessage(&pfrom, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::WTXIDRELAY));
```
> I'm slightly confused about why this was originally sending `INIT_PROTO_VERSION` (and below).

I don't know the whole historical context, but it seems the reason is that on that stage the `INIT_PROTO_VERSION` is the only known version on which both the node and its peer are agree as version is changed later: https://github.com/bitcoin/bitcoin/blob/4f229d8904f8e3494cab30d61df9f11f1cc06388/src/net_processing.cpp#L2432-L2434

>  I think this means a small change in the p2p message right?

See: https://github.com/bitcoin/bitcoin/pull/17785#issuecomment-568043011

jnewbery replied:
I believe the only object that is serialized differently based on p2p version is a CAddress, so the addresses in the `version` message don't have timestamps.

Changing the serialization version for a wtxid message has no effect.

amitiuttarwar replied:
ah, okay, that clears up my confusion about possible change. thank you!

still doesn't make sense why it was that way,
>  it seems the reason is that on that stage the INIT_PROTO_VERSION is the only known version on which both the node and its peer are agree as version is changed later:

yeah, but the local `nSendVersion` was set earlier

but I don't think that's super important now that I understand the context better. 

thanks!

jnewbery commented:
I think `nVersion` is still useful for `getpeerstat` to display which version was received from the peer, although perhaps it could be renamed to `m_version_received` or similar in a future PR.

fjahr commented:
Code review ACK ddefb5c0b759950942ac03f28c43b548af7b4033

Nice cleanup!

amitiuttarwar commented:
code review but untested ACK ddefb5c0b7


------

for another PR: 

@hebasto 
> @jnewbery raised the same question. This could be resolved in a follow-up PR.
I think we need smth like bool CNode::VersionMsgAlreadyReceived().

yeah, but I got slightly confused because based on [this comment](https://github.com/bitcoin/bitcoin/pull/17785#issuecomment-679416590) I thought you might have been getting rid of `nVersion` entirely. no worries :) I just wanted to highlight for future improvements.

MarcoFalke commented:
Post merge ACK, but would be good to remove nVersion from the tests as well in the last commit. See #20131

MarcoFalke commented on src/net.cpp:
```diff
@@ -621,32 +621,6 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete
-    // once a version message has been successfully processed. Any attempt to
-    // set this twice is an error.
-    if (nSendVersion != 0) {
-        error("Send version already set for node: %i. Refusing to change from %i to %i", id, nSendVersion, nVersionIn);
```
Any reason to remove this debug log? I think it would be good to keep this to avoid regressing on this in the future.

MarcoFalke replied:
Fixed in #20138
